### ISSUE TITLE
Problem in webpack

### ISSUE THREAD

**Issue by 23suhail88** [wigahluk/play-webpack#1]
I am facing issue while running activator dist. below is the stack trace of the same. can you please suggest something for the same.

C:\Workspaces\HorizenClientConfigTool_POC\play-webpack-master>activator dist
[info] Loading project definition from C:\Workspaces\HorizenClientConfigTool_POC
\play-webpack-master\project
[info] Set current project to play-webpack (in build file:/C:/Workspaces/Horizen
ClientConfigTool_POC/play-webpack-master/)
java.io.IOException: Cannot run program "npm" (in directory ".\app\frontend"): C
reateProcess error=2, The system cannot find the file specified
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
        at sbt.SimpleProcessBuilder.run(ProcessImpl.scala:349)
        at sbt.AbstractProcessBuilder.run(ProcessImpl.scala:126)
        at sbt.AbstractProcessBuilder.$bang(ProcessImpl.scala:154)
        at $1fb2c49e205bd9c307d1$$anonfun$$sbtdef$1.apply$mcV$sp(C:\Workspaces\H
orizenClientConfigTool_POC\play-webpack-master\build.sbt:31)
        at $1fb2c49e205bd9c307d1$$anonfun$$sbtdef$1.apply(C:\Workspaces\HorizenC
lientConfigTool_POC\play-webpack-master\build.sbt:31)
        at $1fb2c49e205bd9c307d1$$anonfun$$sbtdef$1.apply(C:\Workspaces\HorizenC
lientConfigTool_POC\play-webpack-master\build.sbt:31)
        at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
        at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
        at sbt.std.Transform$$anon$4.work(System.scala:63)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:22
6)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:22
6)
        at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
        at sbt.Execute.work(Execute.scala:235)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
        at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestric
tions.scala:159)
        at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:51
1)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.
java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor
.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: CreateProcess error=2, The system cannot find th
e file specified
        at java.lang.ProcessImpl.create(Native Method)
        at java.lang.ProcessImpl.<init>(ProcessImpl.java:386)
        at java.lang.ProcessImpl.start(ProcessImpl.java:137)
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
        at sbt.SimpleProcessBuilder.run(ProcessImpl.scala:349)
        at sbt.AbstractProcessBuilder.run(ProcessImpl.scala:126)
        at sbt.AbstractProcessBuilder.$bang(ProcessImpl.scala:154)
        at $1fb2c49e205bd9c307d1$$anonfun$$sbtdef$1.apply$mcV$sp(C:\Workspaces\H
orizenClientConfigTool_POC\play-webpack-master\build.sbt:31)
        at $1fb2c49e205bd9c307d1$$anonfun$$sbtdef$1.apply(C:\Workspaces\HorizenC
lientConfigTool_POC\play-webpack-master\build.sbt:31)
        at $1fb2c49e205bd9c307d1$$anonfun$$sbtdef$1.apply(C:\Workspaces\HorizenC
lientConfigTool_POC\play-webpack-master\build.sbt:31)
        at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
        at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
        at sbt.std.Transform$$anon$4.work(System.scala:63)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:22
6)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:22
6)
        at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
        at sbt.Execute.work(Execute.scala:235)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
        at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestric
tions.scala:159)
        at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:51
1)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.
java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor
.java:617)
        at java.lang.Thread.run(Thread.java:745)
[error](*:webpackBuild) java.io.IOException: Cannot run program "npm" (in direc
tory ".\app\frontend"): CreateProcess error=2, The system cannot find the file s
pecified
[error] Total time: 0 s, completed Jun 17, 2016 12:44:14 PM

**Comment by setrar** [comment]
https://github.com/nouhoum/play-react-webpack/issues/2#issuecomment-248637763

**Comment by JFCote** [comment]
I have the same problem but using "activator run". It seems that "npm" needs to be in the ./app/frontend folder... But even then, the Process function in the scala code fails...

Since I don't understand Scala code, I'm stuck right now...

Unfortunately, as far as I know, either this project is not working as is, or there is something missing in the documentation.

@wigahluk , any idea what is going on?

**Comment by wigahluk** [comment]
Are you running on Windows?

**Comment by JFCote** [comment]
@wigahluk : yes, windows 10

**Comment by wigahluk** [comment]
@JFCote thanks. I don't have a windows machine at hand to test. I'll try to install a vm and try.

**Comment by wigahluk** [comment]
@JFCote I haven't had the chance to ran it but I think the problem is with `build.sbt` file, lines 31 and 37. I think that if you change the `/` character with `\` it should work. I don't know if the dot `.` will behave the same, so you may also need to change `"./app/frontend"` to `"app\frontend"`.

**Comment by JFCote** [comment]
ok I will try and let you know! Thanks!

**Comment by JFCote** [comment]
Ok, here is what is missing to make it work in Windows 10:

In `npm.scala`, the `afterStarted` function needs to look like this:

```
override def afterStarted(addr: InetSocketAddress): Unit = {
  if (System.getProperty("os.name").toUpperCase().indexOf("WIN") >= 0) {
    process = Option(Process("cmd /c npm start", base).run)
  } else {
    process = Option(Process("npm start", base).run)
  }
}
```

And build.sbt must look like this:

```
import play.sbt.PlayImport.PlayKeys.playRunHooks

name := """play-webpack"""

version := "1.0-SNAPSHOT"

lazy val root = (project in file(".")).enablePlugins(PlayScala)

scalaVersion := "2.11.6"

libraryDependencies ++= Seq(
    cache,
    ws,
    specs2 % Test
)

// Starts: Prevent documentation of API for production bundles
sources in (Compile, doc) := Seq.empty
publishArtifact in (Compile, packageDoc) := false
// Ends.


resolvers += "scalaz-bintray" at "http://dl.bintray.com/scalaz/releases"

// Play provides two styles of routers, one expects its actions to be injected, the
// other, legacy style, accesses its actions statically.
routesGenerator := InjectedRoutesGenerator

// Starts: Webpack build task
val webpackBuild = taskKey[Unit]("Webpack build task.")
webpackBuild := {
  if (System.getProperty("os.name").toUpperCase().indexOf("WIN") >= 0) {
    Process("cmd /c npm run build", file("\\app\\frontend")).run
  } else {
    Process("npm run build", file("\\app\\frontend")).run
  }
}
(packageBin in Universal) <<= (packageBin in Universal) dependsOn webpackBuild
// Ends.


// Starts: Webpack server process when running locally and build actions for productionbundle
lazy val frontendDirectory = baseDirectory {_ / "app\\frontend"}
playRunHooks <+= frontendDirectory.map(base => WebpackServer(base))
// Ends.

```

**Comment by wigahluk** [comment]
@JFCote I just did a push with your changes. Thanks!

**Comment by JFCote** [comment]
@wigahluk : No problem! Thanks!

**Comment by jericmason** [comment]
I've made a few small changes to the project to get it up-to-date with newer play versions. I've updated sbt-plugin to 2.5.15 and sbt.version to 0.13.13. I use sbt rather than activator to build/run and it builds successfully; however, when I run the project I end up with this error:

`> play-webpack@1.0.0 start C:\Projects\play-webpack\app\frontend
> node ./webpack/webpack.server

module.js:327
    throw err;
    ^

Error: Cannot find module 'webpack'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (C:\Projects\play-webpack\app\frontend\webpack\webpack.server.js:5:15)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Function.Module.runMain (module.js:441:10)

npm ERR! Windows_NT 10.0.15063
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "start"
npm ERR! node v4.5.0
npm ERR! npm  v2.15.9
npm ERR! code ELIFECYCLE
npm ERR! play-webpack@1.0.0 start: `node ./webpack/webpack.server`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the play-webpack@1.0.0 start script 'node ./webpack/webpack.server'.
npm ERR! This is most likely a problem with the play-webpack package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./webpack/webpack.server
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs play-webpack
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!
npm ERR!     npm owner ls play-webpack
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Projects\play-webpack\app\frontend\npm-debug.log`

I'm currently running on Windows 10. Any ideas about what might be going wrong here?

**Comment by wigahluk** [comment]
@jericmason
I'ld try with Node 5 or newer. I'm not able to reproduce the same error that you see but I'm running on a different OS and different versions of Node, npm, SBT and Scala.

Also,I have made several upgrades to the project, which now is using `sbt` instead of `Activator` thanks to @benmccann. 

I found some issues after upgrading my environment to Node 8 because of the new `package-lock.json` behavior. For me, removing the lock file before installing new dependencies or before running `npm install` solved my issues.

I'm closing this as `Activator` is no longer used.
